### PR TITLE
net/udp: Relay full UdpMetadata instead of only remote endpoint in poll_ functions

### PR DESCRIPTION
**Pull request by chrysn**
This is a breaking change for users of the poll_ functions. (Some might not notice if they already pass in an IpEndpoint into poll_send_to, or discard that item in poll_recv_from).

It enables embassy users to utilize enhancements in smoltcp that are hopefully coming up in https://github.com/smoltcp-rs/smoltcp/pull/904. Note that those two do not need to lockstep: We can change this now (without gaining many benefits, just access to the packet ID), and once smoltcp updates, the local address will be available as well.

This is the minimal change from https://github.com/embassy-rs/embassy/pull/2519, and enables that to progress more easily -- once this one is in, the larger PR should even be testable without adding too many `[patches.crates-io]` to already complex setups.

[edit: simplified link]

### CODE REVIEW

File: embassy-net/src/udp.rs [hunk 1]
```diff
@@ -122,9 +124,13 @@ impl<'a> UdpSocket<'a> {
     ///
     /// When a datagram is received, this method will return `Poll::Ready` with the
     /// number of bytes received and the remote endpoint.
-    pub fn poll_recv_from(&self, buf: &mut [u8], cx: &mut Context<'_>) -> Poll<Result<(usize, IpEndpoint), RecvError>> {
+    pub fn poll_recv_from(
+        &self,
+        buf: &mut [u8],
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<(usize, UdpMetadata), RecvError>> {
```
Comment (Dirbaio) [inline embassy-net/src/udp.rs:131]: `recv_from` and `poll_recv_from` should return the same. can you change `recv_from` too?

(the convention is `poll_*` functions are identical to their non-poll counterpart, except manually polled)

Comment (chrysn) [reply embassy-net/src/udp.rs:131]: Done in a fixup; this even makes the total diff smaller. Shall I squash and force-push now or when reviews are done?

**Comment by chrysn** [comment]
The merge of main into the branch (AIU this is the pattern used in this repo) fixed the nightly builds. The builds are fixed by providing a fixup as requested in https://github.com/embassy-rs/embassy/pull/2790#pullrequestreview-1999747527 also for send_to, so that send_to is now consistent with poll_send_to, and another one that fixes unused import warnings.

**Review by Dirbaio — APPROVED** [review]
lgtm, thakns!

**Pull request merged**